generators/csharp/sdk/src/oauth/OauthTokenProviderGenerator.ts

```diff
@@ -0,0 +1,267 @@
+            isAsync: true,
+            name: "GetAccessTokenAsync",
+            parameters: [],
+            body: this.getAccessTokenBody(),
+            return_: csharp.Type.string()
+        });
+    }
+
+    private getAccessTokenBody(): csharp.CodeBlock {
+        const expiresIn = this.getExpiresIn();
+        const tokenEndpoint = this.scheme.configuration.tokenEndpoint;
+
+        return csharp.codeblock((writer) => {
+            writer.controlFlow(
+                "if",
+                csharp.codeblock((writer) => {
+                    writer.write(`${OauthTokenProviderGenerator.ACCESS_TOKEN_FIELD.name} == null`);
+                    // check expiresIn if present in the IR
+                    if (expiresIn != null) {
+                        writer.write(`|| DateTime.UtcNow >= ${OauthTokenProviderGenerator.EXPIRES_AT_FIELD.name}`);
+                    }
+                })
+            );
+
+            writer.writeNodeStatement(
+                csharp.codeblock((writer) => {
+                    const requestType = this.tokenEndpoint.requestBody?._visit({
+                        reference: (value) => {
+                            return this.context.csharpTypeMapper.convert({ reference: value.requestBodyType });
+                        },
+                        inlinedRequestBody: (value) => {
+                            return csharp.Type.reference(
+                                this.context.getRequestWrapperReference(
+                                    this.tokenEndpointReference.serviceId,
+                                    value.name
+                                )
+                            );
+                        },
+                        fileUpload: (value) => undefined,
+                        bytes: (value) => undefined,
+                        _other: (value) => undefined
+                    });
+                    if (requestType == null || requestType.internalType.type !== "reference") {
+                        throw new Error("Failed to get request class reference");
+                    }
+
+                    writer.write(`var tokenResponse = `);
+                    writer.writeNode(
+                        csharp.invokeMethod({
+                            async: true,
+                            method: `client.${this.tokenEndpoint.name.pascalCase.safeName}`,
+                            arguments_: [
+                                csharp.instantiateClass({
+                                    classReference: requestType.internalType.value,
+                                    // TODO(dsinghvi): assumes only top level client id and client secret inputs
+                                    arguments_: [
+                                        {
+                                            name: this.scheme.configuration.tokenEndpoint.requestProperties.clientId
+                                                .property.name.name.snakeCase.safeName,
```

Comment:
Is there a helper we can use for the naming here? I don't think `snakeCase` is right.

Reply:
yup definitely